ls1intum/Artemis
# `Programming exercises`: Add access tokens to interact with Gitlab.

### Checklist
#### General


#### Server
#### Client
#### Changes affecting Programming Exercises

### Motivation and Context
When using Artemis with a SAML/Shibboleth setup, two options regarding the user management on the VCS (Gitlab) exist:
1. Use SAML only to create an Artemis user with a separate password. This password can the be used to access Gitlab. This may not be desired as the user has to manage two passwords: The password of the SAML user and the password of the Artemis user.
2. Use SAML to login to Artemis, and use SAML to login to Gitlab. Gitlab does not know the password of the user in the IdP and creates an internal one (which is never used be the user).

This PR focuses on the second option:
As Gitlab does not know the password of the user, it creates a random one (`internal password`). This password is not known to the user as all logins happen using SAML.
This is no problem when the Gitlab Web UI is used (as here the login via SAML is possible), however when cloning, there are two options:
1. Use SSH. The user must upload an SSH Key first in Gitlab.
2. Use HTTPS. The user must authenticate using the `internal password` (which is unknown).

As option 1 is not trivial for new users, we want to use option 2, however the `internal password` is unknown (and can not be extracted by Artemis).
This PR adds an `access token` that can be used instead of the `internal password` when interacting with the git repository.


### Description
The `access token` is created when the user is initially created in Gitlab by Artemis and stored in the database.
It is also created (if it does not exist) when starting a programming exercise.
When the `Clone dialog` is shown, the token is inserted in the URL, so that the user can simply copy the URL and clone directly.
To prevent leaking the token by accident, the token is replaced by `**********` in the dialog and only inserted when using the `Copy URL` button (or the `Clone in SourceTree` button).

The config option `artemis.version-control.version-control-access-token` must be set to `true` to use this feature.

### Steps for Testing

Prerequisites:
- Gitlab setup
- `artemis.version-control.version-control-access-token` is set to `true`
- 1 programming exercise

I've performed the required config changes on TS2.

1. As admin: Create a new user in Artemis (e.g. as Administrator, or by registering).
2. As admin/user: Add the user to the course with the programming exercise
3. As user: Start the programming exercise
4. As user: Open the clone dialog. It should contain the token replaced with `**********`
5. As user: Copy the URL. It should contain the actual token (instead of the placeholder)
6. As user: Clone the repository using the URL, make some changes, and push. This should all be possible without specifying a password.

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage






### Screenshots
![image](https://user-images.githubusercontent.com/5084100/156181158-ae8bcd5b-2660-4735-a54e-c8a0137adef8.png)
![Peek 2022-03-01 14-53](https://user-images.githubusercontent.com/5084100/156181861-3be4260a-3fa3-4281-bef4-a5c8484b2c84.gif)
